tests/test_protocol_object_conversion.py

```diff
@@ -6,62 +8,119 @@
+                API_KEY = 0
+                API_VERSION = 0
+                SCHEMA = Schema(("myobject", Int16))
 
         tc = TestClass(myobject=0)
         assert tc.get_item("myobject") == 0
         with pytest.raises(KeyError):
             tc.get_item("does-not-exist")
 
-    def test_with_empty_schema(self, superclass):
-        class TestClass(superclass):
-            API_KEY = 0
-            API_VERSION = 0
-            RESPONSE_TYPE = None  # To satisfy the Request ABC
-            SCHEMA = Schema()
+    def test_with_empty_schema(
+        self, superclass: Type[Union[Request, Response]]
+    ) -> None:
+        if issubclass(superclass, Request):
+
+            class TestClass(Request):
+                API_KEY = 0
+                API_VERSION = 0
+                RESPONSE_TYPE = Response
+                SCHEMA = Schema()
+        else:
+
+            class TestClass(Response):  # type: ignore[no-redef]
+                API_KEY = 0
+                API_VERSION = 0
+                SCHEMA = Schema()
 
         tc = TestClass()
         tc.encode()
         assert tc.to_object() == {}
 
-    def test_with_basic_schema(self, superclass):
-        class TestClass(superclass):
-            API_KEY = 0
-            API_VERSION = 0
-            RESPONSE_TYPE = None  # To satisfy the Request ABC
-            SCHEMA = Schema(("myobject", Int16))
+    def test_with_basic_schema(
+        self, superclass: Type[Union[Request, Response]]
+    ) -> None:
+        if issubclass(superclass, Request):
+
+            class TestClass(Request):
+                API_KEY = 0
+                API_VERSION = 0
+                RESPONSE_TYPE = Response  # To satisfy the Request ABC
+                SCHEMA = Schema(("myobject", Int16))
+        else:
+
+            class TestClass(Response):  # type: ignore[no-redef]
+                API_KEY = 0
+                API_VERSION = 0
+                SCHEMA = Schema(("myobject", Int16))
 
         tc = TestClass(myobject=0)
         tc.encode()
         assert tc.to_object() == {"myobject": 0}
 
-    def test_with_basic_array_schema(self, superclass):
-        class TestClass(superclass):
-            API_KEY = 0
-            API_VERSION = 0
-            RESPONSE_TYPE = None  # To satisfy the Request ABC
-            SCHEMA = Schema(("myarray", Array(Int16)))
+    def test_with_basic_array_schema(
+        self, superclass: Type[Union[Request, Response]]
+    ) -> None:
+        if issubclass(superclass, Request):
+
+            class TestClass(Request):
+                API_KEY = 0
+                API_VERSION = 0
+                RESPONSE_TYPE = Response  # To satisfy the Request ABC
+                SCHEMA = Schema(("myarray", Array(Int16)))
+        else:
+
+            class TestClass(Response):  # type: ignore[no-redef]
+                API_KEY = 0
+                API_VERSION = 0
+                SCHEMA = Schema(("myarray", Array(Int16)))
```

Comment:
Looks like the same code is repeated a lot of times. I think we can put it into generic function like this:
```python
def envelope_with_schema(superclass: ET, schema: Schema) -> ET:
    ...
```
Or even a parametrized factory fixture with single `schema` argument.